### ISSUE TITLE
multi-factor authentication not supported?

### ISSUE THREAD

**Issue by ShawnTalbert** [microsoft/azure-devops-intellij#239]
* want to check out a repo via "Azure Devops Git" menu
* connect using the "Sign In..."
* I enter email address, and taken to organizational signin page (username/password)
* login succeeds but _I am never challenged for multi-factor authentication_
* Result - login appears successful but I get an empty repository list, making this checkout feature useless.

My organization **requires** MFA for access to Azure DevOps repos.

How can I get the plugin's login flow to include the MFA challenge like the normal web browser flow does?

Any workarounds?

**Comment by ForNeVeR** [comment]
Thanks for the report. For now, could you please check if the Device flow authentication will work for you?

1. Open IDE **Settings → Version Control → Azure DevOps Services / TFS**
2. Select **Authentication Method**: **Device flow**:
   ![image](https://user-images.githubusercontent.com/92793/63567816-3b7b8980-c59d-11e9-9217-b1995d1683a3.png)

**Comment by ForNeVeR** [comment]
@ShawnTalbert, this is unusual. Are you aware of any organization settings that force that? I was unable to find this particular message in Google at all. I've also tried to find any related settings in my Azure organization settings to no avail.

**Comment by ShawnTalbert** [comment]
I asked and IT doesn't think there's setting to prevent it, but admittedly our IT has several layers depth to it.

In any case, if 'device flow' = 'trust my entire laptop' I'm not sure that's wise from a security perspective?

Does that exception from inside the plugin help at all? Is there hope to expand the plugin to support MFA for individual users?

**Comment by ForNeVeR** [comment]
Could you please tell more details about your account? Is it a corporate one (i.e. not a Microsoft one)? How does it work?

I've set up 2FA for my own account and I was able to successfully log into Azure DevOps using the plugin, so I'd say that technically there's nothing preventing us from supporting it (though I wasn't able to force MFA usage for any of my Azure DevOps organizations).

**Comment by ForNeVeR** [comment]
Also, does Visual Studio work for you? I've found [a report](https://developercommunity.visualstudio.com/content/problem/403750/connect-to-azure-devops-in-b2b-aad-with-mfa.html) that VS' authentication fails for a similar case, too.

**Comment by ShawnTalbert** [comment]
My account is a corporate one. Clicking 'Connect to Azure' from the VS Start Page - the browser popup takes me to my 'organization' login page, I'm prompted for initial password credential, not prompted for MFA, then nothing happens... but the output panel (Note I'm on VS 2015) displays this text:

> We were unable to automatically populate your Visual Studio Team Services accounts.
    The following error was encountered: VS30063: You are not authorized to access https://app.vssps.visualstudio.com.

**Comment by ForNeVeR** [comment]
Okay, thanks for your report.

It looks like MFA for corporate accounts works very differently from 2FA on Microsoft ones. We have plans to use [microsoft-authentication-library-for-java](https://github.com/AzureAD/microsoft-authentication-library-for-java) which claims to support MFA.

How does your corporate authentication work? You sign in to your computer using the same credentials that will be used for resource access? I.e. are we supposed to use the authentication details from the used currently signed in? Could you please tell me a couple of buzzwords about your authentication scheme, so I'd ask local wise people about it? How does it organized? Where are the accounts hosted?

**Comment by ShawnTalbert** [comment]
My corporate account is sync’d to Azure AD via the AD Connect tool. Then that account was added to the AppDev Azure Ad Tennant as a Guest Account. An Azure AD P2 license was assigned and we are enforcing MFA for all guest users.

I hope MSAL For Java can save the day!

**Comment by roelver** [comment]
I have a similar issue Signing in to Azure Devops in the IntelliJ plugin with my company AD account. 
I tried with both userid/password and Device-flow option.
For me, the authentication procedure is working as usual:
- Login with company account-id
- Forwarded to a page where I need to enter my password
- Forwarded to a page where I have to enter the code that I received by a text message on my cell phone
So far so good and the authentication window is closed.
But after this, I get an error on the plugin's Sign-in window: ' Unable to Authenticate to Azure DevOps Services. Verify your network connection and credentials and try "Sign in..." again'.

See attached stack trace. Same error when trying the Device-flow option.

[IntelliJ_AzureDevOps_Signin_error.txt](https://github.com/microsoft/azure-devops-intellij/files/3659159/IntelliJ_AzureDevOps_Signin_error.txt)

**Comment by faustool** [comment]
When I try, I get the exception below. I am behind the network proxy, but wasn't prompted for proxy authentication by the IDE as usual.

```
getAuthenticationInfoAsync failed

com.microsoft.alm.client.model.VssResourceNotFoundException: API resource location f83735dc-483f-4238-a291-d45f6080a9af is not registered on https://app.vssps.visualstudio.com. java.net.UnknownHostException: No such host is known (app.vssps.visualstudio.com)
	at com.microsoft.alm.client.AlmHttpClientBase.createTarget(AlmHttpClientBase.java:138)
	at com.microsoft.alm.client.AlmHttpClientBase.createRequest(AlmHttpClientBase.java:463)
	at com.microsoft.alm.client.AlmHttpClientBase.createRequest(AlmHttpClientBase.java:334)
	at com.microsoft.visualstudio.services.account.AccountHttpClient.getMyProfile(AccountHttpClient.java:49)
	at com.microsoft.alm.plugin.authentication.facades.VsoAuthInfoProvider.getAuthenticationInfo(VsoAuthInfoProvider.java:123)
	at com.microsoft.alm.plugin.authentication.facades.VsoAuthInfoProvider.getAuthenticationInfoAsync(VsoAuthInfoProvider.java:95)
	at com.microsoft.alm.plugin.authentication.VsoAuthenticationProvider.authenticateAsync(VsoAuthenticationProvider.java:69)
	at com.microsoft.alm.plugin.authentication.AuthHelper.getAuthenticationInfoSynchronously(AuthHelper.java:59)
	at com.microsoft.alm.plugin.context.ServerContextManager.getAuthenticationInfo(ServerContextManager.java:480)
	at com.microsoft.alm.plugin.context.ServerContextManager.createContextFromGitRemoteUrl(ServerContextManager.java:364)
	at com.microsoft.alm.plugin.context.ServerContextManager.createContextFromGitRemoteUrl(ServerContextManager.java:345)
	at com.microsoft.alm.plugin.context.ServerContextManager.getAuthenticatedContext(ServerContextManager.java:325)
	at com.microsoft.alm.plugin.context.ServerContextManager.getUpdatedContext(ServerContextManager.java:563)
	at com.microsoft.alm.plugin.operations.Operation$1.run(Operation.java:184)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: javax.ws.rs.ProcessingException: java.net.UnknownHostException: No such host is known (app.vssps.visualstudio.com)
	at org.glassfish.jersey.apache.connector.ApacheConnector.apply(ApacheConnector.java:472)
	at org.glassfish.jersey.apache.connector.ApacheConnector.apply(ApacheConnector.java:479)
	at org.glassfish.jersey.client.ClientRuntime.lambda$null$3(ClientRuntime.java:163)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:288)
	at org.glassfish.jersey.client.ClientRuntime.lambda$createRunnableForAsyncProcessing$4(ClientRuntime.java:139)
	... 5 more
Caused by: java.net.UnknownHostException: No such host is known (app.vssps.visualstudio.com)
	at java.base/java.net.Inet4AddressImpl.lookupAllHostAddr(Native Method)
	at java.base/java.net.InetAddress$PlatformNameService.lookupAllHostAddr(InetAddress.java:929)
	at java.base/java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1515)
	at java.base/java.net.InetAddress$NameServiceAddresses.get(InetAddress.java:848)
	at java.base/java.net.InetAddress.getAllByName0(InetAddress.java:1505)
	at java.base/java.net.InetAddress.getAllByName(InetAddress.java:1364)
	at java.base/java.net.InetAddress.getAllByName(InetAddress.java:1298)
	at org.apache.http.impl.conn.SystemDefaultDnsResolver.resolve(SystemDefaultDnsResolver.java:45)
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:111)
	at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:353)
	at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:380)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:71)
	at org.glassfish.jersey.apache.connector.ApacheConnector.apply(ApacheConnector.java:426)
	... 14 more
```

**Comment by ForNeVeR** [comment]
@faustool, your case may be a duplicate of #89. Could you try to use device flow auth instead of the default one? It could help since it uses your system browser.

**Comment by faustool** [comment]
Hi.

I did try that and the error is the same.

**Comment by ForNeVeR** [comment]
Depends on #301.
